File: BOJ/27_동적 계획법과 최단거리 역추적/11779_최소비용구하기2.cpp
#include <iostream>
#include <string.h>
#include <vector>
#include <queue>

/** 
 * 같은 방식으로 푸는데 path 출력 부분에서 오답이 되어버린다.
 * 왜 구럴까
**/

#define MAXN 100000
#define INF 987654321
using namespace std;

vector<pair<int,int> > adjs[MAXN+1];
int dist[MAXN+1],par[MAXN+1];
int N,M,x,y,c,st,ed;

int getLen(string s) {
    int ret = 0;
    for(int i=0; i<s.length(); i++) {
        if(s[i]== ' ') continue;
        ret += 1;
    }
    return ret;
}

void dji() {

    // priority_queue<pair<int,pair<int,string> > > pq; // -cost, here, path
    // pq.push(make_pair(0,make_pair(st,to_string(st))));
    priority_queue<pair<int,int> > pq; // -cost , here
    pq.push(make_pair(0,st));
    dist[st] = 0;
    while (!pq.empty()) {
        int cost = -pq.top().first;
        int here = pq.top().second;
        // string path = pq.top().second.second;
        pq.pop();
        // if(here == ed) {
        //     cout << cost << "\n" << getLen(path) << "\n" << path << "\n";
        //     return ;
        // }
        if(dist[here] < cost) continue;

        for(int i=0; i<adjs[here].size(); i++) {
            int next = adjs[here][i].first;
            int nextCost = cost + adjs[here][i].second;

            if(dist[next] > nextCost) {
                dist[next] = nextCost;
                pq.push(make_pair(-nextCost,next));
                par[next] = here;
                // pq.push(make_pair(-nextCost,make_pair(next,path+" "+to_string(next))));
            }
        }
    }
    vector<int> path;
    int tmp = ed;
    path.push_back(ed);
    while(st != tmp) {
        path.push_back(par[tmp]);
        tmp = par[tmp];
    }
    cout << dist[ed] << "\n" << path.size() << "\n";
    for(int i=path.size()-1; i>=0; i--) cout << path[i] << " ";
    cout << "\n";
}


int main() {

    cin >> N >> M;
    for(int i=0; i<=MAXN; i++) { dist[i] = INF; par[i] = 0; }
    for(int i=0; i<M; i++) {
        cin >> x >> y >> c;
        adjs[x].push_back(make_pair(y,c));
    }
    cin >> st >> ed;
    dji();
    return 0;
}

File: BOJ/27_동적 계획법과 최단거리 역추적/11779_최소비용구하기2.py
import heapq
import sys

read = sys.stdin.readline
INF = int(1e9)

# DSLR 과 동일한 풀이과정

def dji() :
    pq = []
    heapq.heappush(pq,[0,st,str(st)]) # cost , here , path
    dp[st] = 0
    while pq :
        cost,here,path = heapq.heappop(pq)
        if here == ed :
            print(cost)
            print(len(path.split()))
            print(path)
            break
        if dp[here] < cost : continue

        for next_,dist in adjs[here] :
            nextCost = cost + dist

            if dp[next_] > nextCost :
                dp[next_] = nextCost
                heapq.heappush(pq,[nextCost,next_,path+" "+str(next_)])

N = int(read())
M = int(read())
adjs = [[] for _ in range(N+1)]
dp = [INF for _ in range(N+1)]
par = [0 for _ in range(N+1)]
for _ in range(M) :
    x,y,c = map(int,read().split())
    adjs[x].append([y,c])

st,ed = map(int,read().split())
dji()